flink/v1.14/flink/src/main/java/org/apache/iceberg/flink/sink/RowDataTaskWriterFactory.java

```diff
@@ -70,9 +72,19 @@ public RowDataTaskWriterFactory(Table table,
     if (equalityFieldIds == null || equalityFieldIds.isEmpty()) {
       this.appenderFactory = new FlinkAppenderFactory(schema, flinkSchema, table.properties(), spec);
     } else {
-      // TODO provide the ability to customize the equality-delete row schema.
-      this.appenderFactory = new FlinkAppenderFactory(schema, flinkSchema, table.properties(), spec,
-          ArrayUtil.toIntArray(equalityFieldIds), schema, null);
+      // In upsert mode, only the new row is emitted using INSERT row kind. Therefore, the deletion schema
+      // must be only the equality fields for that the writer can use the new row to issue a delete by key.
+      //
+      // Outside the upsert case, the entire deleted row is emitted as UPDATE_BEFORE or DELETE row kind,
+      // so the deletion schema must use the whole rows schema in case elements are deleted based on additional
+      // fields.
+      if (upsert) {
```

Comment:
You should be able to avoid changing the `else` case by adding this as an `else if`, right? That would reduce the risk of conflict.

Reply:
Fair point (assuming you mean the risk of conflict for people who are cherry-picking).

Updating.